### ISSUE TITLE
Compute renderer displays garbage at high IRs

### ISSUE THREAD

**Issue by FireNX70** [melonDS-emu/melonDS#2047]
Resolutions higher than 6x are very prone to displaying garbage. 6x will display garbage in certain spots, although Zelda Spirit Tracks triggers this pretty easily. 5x seems fine. Tested on a 1080Ti + Win10. Tested with https://github.com/melonDS-emu/melonDS/commit/a72b79a55ad2d61811af11b1b911f6af863f66c2, so https://github.com/melonDS-emu/melonDS/commit/5df83c97c766bff3da8ba5a1504a6a5974467133 doesn't fix it.

Some examples
<img width="424" alt="pkmn_bw_6x" src="https://github.com/melonDS-emu/melonDS/assets/23368648/1db1a9f9-2a0e-44f9-95a4-777280cde169"> Pokemon BW Driftveil Bridge 6x
<img width="424" alt="pkmn_pt_spot_5x" src="https://github.com/melonDS-emu/melonDS/assets/23368648/8ab57000-33a6-4769-855f-3f61597c73f9"> Pokemon Platinum 5x
<img width="423" alt="pkmn_pt_spot_6x" src="https://github.com/melonDS-emu/melonDS/assets/23368648/940e17a0-47ec-446c-b504-af6b085c8977"> Pokemon Platinum 6x
<img width="424" alt="pkmn_pt_next_to_spot_6x" src="https://github.com/melonDS-emu/melonDS/assets/23368648/cf890204-9381-4690-98fb-4291f6fdfc14"> Pokemon Platinum 6x
<img width="422" alt="pkmn_pt_next_to_spot_7x" src="https://github.com/melonDS-emu/melonDS/assets/23368648/8d8b220b-de9b-4044-9844-91c310cc4af7"> Pokemon Platinum 7x
<img width="421" alt="tloz_ph_6x" src="https://github.com/melonDS-emu/melonDS/assets/23368648/8c922d04-1289-4e43-8096-b8736481e77c"> Zelda Phantom Hourglass 6x
<img width="423" alt="tloz_ph_7x" src="https://github.com/melonDS-emu/melonDS/assets/23368648/61409057-4428-4658-a4de-00b630d09f7b"> Zelda Phantom Hourglass 7x
<img width="423" alt="tloz_st_5x" src="https://github.com/melonDS-emu/melonDS/assets/23368648/88125b21-d121-4416-b7c4-96f374aa1fe7"> Zelda Spirit Tracks 5x
<img width="423" alt="tloz_st_6x" src="https://github.com/melonDS-emu/melonDS/assets/23368648/d02931ff-88ca-43dd-a299-270cef0889ce"> Zelda Spirit Tracks 6x

**Comment by FireNX70** [comment]
I tried doubling MaxWorkTiles and it didn't fix it. Increasing the size of FinalTileMemory didn't do anything either.

**Comment by takeshineale128** [comment]
no issues on my gtx 1660 gpu at 16x ir on win 11 with driver version v552.44

**Comment by FireNX70** [comment]
I've tested both updating the driver and using a T1000. Neither fixed this. So it's not the driver version or the architecture. I really don't think the OS would make a difference, especially if it's Win10 vs Win11. Were you really testing with the compute renderer?

**Comment by takeshineale128** [comment]
opengl

**Comment by FireNX70** [comment]
There's two OpenGL renderers. Are you testing with a dev version or are you using 0.9.5?

**Comment by takeshineale128** [comment]
0.9.5

**Comment by RSDuck** [comment]
> Increasing the size of FinalTileMemory didn't do anything either.

FinalTileMemory isn't dynamically allocated, it cannot overflow.

Not really sure what could be the issue here. It seems to be definitely tile related and the index should not overflow (yet).

**Comment by nadiaholmquist** [comment]
Can't seem to reproduce this on an AMD GPU (Steam Deck). Will try on other stuff later.

**Comment by FireNX70** [comment]
@nadiaholmquist I've tested with a Vega 8 (3500U) on Win10 with an old driver from 2023 and it works fine even at 16x, although attempting to switch resolutions while the game was running completely froze the renderer. I've also tested with LLVMpipe in a VM and it also works fine even at 16x. It looks like it's a bug specific to Nvidia. I'll test Nvidia on Linux with the proprietary driver, I expect that it will behave just like it does in Win10.

**Comment by RSDuck** [comment]
oh no, there still might be undefined behaviour somewhere in the shaders. I'm going to run the shaders with mesa and a debug context eventually to see whether it warns about something.

**Comment by takeshineale128** [comment]
not useing dev builds only the 0.9.5 build has no opengl option for classic or compute shader 
![Screenshot 2024-05-18 051021](https://github.com/melonDS-emu/melonDS/assets/5340068/de8e4511-cd58-47e5-a2d2-7a34b1139299)

**Comment by KostaSaizo7** [comment]
Blue Dragon - Awakened Shadow
Windows 10
Nvidia GPU
Compute shader x8 (x7 and below is fine)
https://www.youtube.com/shorts/TX5Nq3RdX5Q

**Comment by FireNX70** [comment]
`glDispatchComputeIndirect` will not log OpenGL errors even if the compute work group counts are too high, and checking these would be annoying since I'd have to read the buffer back from VRAM. However, RenderDoc did capture the values used for these. At 6x none of the calls ever go above the minimum of 65535. At 7x there is one call that does exceed that minimum (`glDispatchComputeIndirect(<1, 1, 79414>)`). This exceeds the maximum supported by my 1080Ti with the 552 driver (it sticks to the minimum of 65535 in the Z axis). I tried replacing the `glDispatchComputeIndirect` call in the loop with `glDispatchCompute(1, 1, 65535)`. This DID have an effect on the garbage, but it did not fix it. It's likely this is the problem on Nvidia. The local sizes all seem fine.

**Comment by RSDuck** [comment]
thank you for debugging that.

Regarding the barriers that is obviously a mistake. Though GL_SHADER_STORAGE_BUFFER is defined as 0x90D2 (which *should* result in a GL error), though if the driver accepts it and only uses the valid bits, most GPUs don't have that fine grained barriers anyway. So there's probably atleast one value set which will insert a correct barrier.

Huh, I always assumed you could do work groups as long as it fits inside a 32-bit int (as it is stored in one). The reason it didn't fix it could be multiple things. If you don't change any of the bindings you will end up just doing the same work multiple times (possibly at the same time).

I need to look into what parallel rdp again, maybe it has a solution to this. I simplified some things, e.g. we always process all polygons at once, because it is limited to a fixed amount on DS. One possible easy out would probably just be to increase the tile size to 16x16 at some point. That would also solve the possible issue of the tile buffer index overrunning.

**Comment by FireNX70** [comment]
> thank you for debugging that.

No problem, once this is fixed I can fully switch to MelonDS.

> Huh, I always assumed you could do work groups as long as it fits inside a 32-bit int (as it is stored in one). The reason it didn't fix it could be multiple things. If you don't change any of the bindings you will end up just doing the same work multiple times (possibly at the same time).

There's a minimum required limit of 65535. Implementations are free to establish higher limits. I haven't checked what limits Mesa and Radeon report, Nvidia has a higher limit on the X dimension but sticks to 65535 for the Y and Z dims.

> I need to look into what parallel rdp again, maybe it has a solution to this. I simplified some things, e.g. we always process all polygons at once, because it is limited to a fixed amount on DS. One possible easy out would probably just be to increase the tile size to 16x16 at some point. That would also solve the possible issue of the tile buffer index overrunning.

I'm not well versed in graphics nor OpenGL, certainly not enough to understand what the compute renderer's doing. Having said that, it looks to me as though the right thing to do would be to cap VariantWorkCount's Z component to the respective GL_MAX_COMPUTE_WORK_GROUP_COUNT and do multiple passes per variant if needed. Easy to say, probably a pain in the ass to implement. I don't think larger tiles would guarantee the work group count wouldn't go above the limit at higher internal resolutions.

**Comment by RSDuck** [comment]
The problem is that the amount of work done per variant is not known to the cpu at the time it issues the indirect compute calls (that's the reason they are indirect in the first place). And letting the GPU do the binning, then waiting on that and then dispatching as much compute jobs as necessary would work, but would just waste CPU time on waiting.

After sleeping over this, there is technically another solution. The other two axis as well and then put all the axis together into one index in the shader.

I quite like the idea of increasing the tile size at very high resolutions though. The increased tile size results in a lower amount of tiles still filling the same area. As long as the total amount of tiles is less than 2^16 no single variant can result in a larger compute job than that.

What of course remains a potential issue is that the tile buffer can overflow, as it has a fixed size. It would be possible to conservely estimate the size necessary each frame by using the bounding boxes of each polygon, but with some polygon shapes that might result in excessive allocations.

Both solutions don't exclude each other, I guess I can also just implement them both.

**Comment by FireNX70** [comment]
Scaling the tile size with the internal resolution sounds good. Do keep in mind there's a limit on local size too, although the minimum of 1024 should be fine (1024x1024 for a single tile would be huge). As for the tile buffer; IMO it's better safe than sorry, especially if it doesn't overestimate by a lot.

**Comment by FireNX70** [comment]
I've been messing around with scaled tiles. 16x16 tiles work fine, 32x32 tiles (the max size, since the tile size is used for the X and Y local sizes in certain shaders and I'm sticking to the default GL_MAX_COMPUTE_WORK_GROUP_INVOCATIONS of 1024) presents a couple of problems.

`glDispatchCompute(TilesPerLine*TileLines/32, 1, 1)`, used for ShaderClearCoarseBinMask, is a problem because the total tile count (TilesPerLine*TileLines) with size 32 tiles will not divide by 32 nicely. I think we divide TilesPerLine * TileLines by 32 because of the shader's local size. That division works fine with both 16 and 48. It won't be optimal for Nvidia (warp size of 32) nor Radeon (RDNA wave size is 32, GCN wave size is 64). I'm not sure if the divider for CoarseBinStride has to match the local size too. If so, this only works if it's 16. It's as far from optimal as 48 for Nvidia and RDNA but probably worse for GCN.

`glDispatchCompute(((gpu.GPU3D.RenderNumPolygons + 31) / 32), ScreenWidth/CoarseTileW, ScreenHeight/CoarseTileH)`, used for ShaderBinCombined, is also a problem. With 32x32 tiles, CoarseTileH would end up being 128. At lower tile sizes it's 32 or 64, which 192 is a multiple of, so it was guaranteed the vertical resolution would be a multiple of CoarseTileH. 192 is not a multiple of 128, so a CoarseTileH of 128 becomes a problem at certain resolutions (specifically, odd multiples of the base resolution). This (I think) is what would cause a blank strip at the bottom of the screen with odd scales. Making CoarseTileCountY variable should fix it, but it introduced more bands (the number of bands matches the scale). With 6-tile tall coarse tiles (when using 32x32 tiles) the height of a coarse tile is 192 (that way we make sure internal resolution is a multiple of CoarseTileH), so the number of coarse tile lines matches the scale. What I think is really happening here is there's actually one two-tile-thick band per coarse tile because there's a hardcoded 4 somewhere so there's 2 rows of tiles that are ignored per coarse tile.

Just in case here's what I've done: https://github.com/FireNX70/melonDS/tree/try-fixing-the-compute-renderer

**Comment by FireNX70** [comment]
Yeah, those are the stripes I was talking about. Starting at 10x it uses 32x32 tiles.

**Comment by FireNX70** [comment]
I think what's happening with the stripes is BinCombined's local_size_x must match CoarseTileCountX * CoarseTileCountY. A quick hack to do that (and changing the associated glDispatchCompute call to add 47 and divide by 48 instead 31 and 32) gets rid of the stripes but causes a bunch of geometry to fail to render.

I've also found (through testing with 16x16 tiles) that the scale at which garbage may start being displayed would be 3.5x (with 8x8 tiles, 7x with 16x16 tiles in practice). The worst case scenario seems to be MKDS' desert track. If the fire balls (which are only there on the last 2 laps) get extremely close to the camera, the scale is high enough and the tile size is small enough; you'll get garbage. Driving through them seems to be the best way to trigger this. The next worst scenario, Pokemon BW Driftveil Bridge, only starts breaking at 5x (8x8) and is fine at 4x (8x8), 7x (16x16) and 8x (16x16).

**Comment by FireNX70** [comment]
I think I got it working. Work cont X just wasn't related to the local size for BinCombined. The only remaining problem is ClearCoarseBinMask, but I think changing the local size to 48 and dividing TilesPerLine*TileLines by 48 should fix that.

**Comment by patataofcourse** [comment]
happens with the melon 1.0 RC on Built to Scale DS (first game in Rhythm Heaven), on 8x upwards

![image](https://github.com/user-attachments/assets/5bbb1d70-f9a3-4f0a-b2be-cdd3fe187ffd)

this is on Windows 10 with an Nvidia RTX 4050

**Comment by HighPingTek** [comment]
Is there any news on any fixes for this issue?

**Comment by user18081972** [comment]
> Is there any news on any fixes for this issue?

Not answering your question directly as I am not part of the melonteam, but you could use my custom build that fixes it: [MelonDS.1.0+NvidiaComputeShaderWorkaround2.0.zip](https://github.com/user-attachments/files/22989829/MelonDS.1.0%2BNvidiaComputeShaderWorkaround2.0.zip) which is based on https://github.com/melonDS-emu/melonDS/pull/2065 and stable version 1.0
